File: internal/game/context.go
package game

import (
	"slices"

	"github.com/google/uuid"
)

type Context struct {
	ActionID          *uuid.UUID  `json:"action_ID"`
	SourcePlayerID    *uuid.UUID  `json:"source_player_ID"`
	ParentActorID     *uuid.UUID  `json:"parent_actor_ID"`
	SourceActorID     *uuid.UUID  `json:"source_actor_ID"`
	TargetActorIDs    []uuid.UUID `json:"target_actor_IDs"`
	TargetPositionIDs []uuid.UUID `json:"target_position_IDs"`
}

func NewContext() Context {
	return Context{
		TargetActorIDs:    []uuid.UUID{},
		TargetPositionIDs: []uuid.UUID{},
	}
}

func WithTargetIDs(context Context, targetActorIDs []uuid.UUID) Context {
	c := context
	c.TargetActorIDs = targetActorIDs
	return c
}

func (g Game) GetTargets(context Context) []Actor {
	targets := g.GetActors(func(a Actor) bool {
		return slices.Contains(context.TargetActorIDs, a.ID)
	})
	posTargets := g.GetActors(func(a Actor) bool {
		if a.PositionID == nil {
			return false
		}

		return slices.Contains(context.TargetPositionIDs, *a.PositionID)
	})

	targets = append(targets, posTargets...)
	return targets
}

func (g Game) GetSource(context Context) (Actor, bool) {
	if context.SourceActorID == nil {
		return Actor{}, false
	}

	return g.GetActorByID(*context.SourceActorID)
}

func (g Game) GetParent(context Context) (Actor, bool) {
	if context.ParentActorID == nil {
		return Actor{}, false
	}

	return g.GetActorByID(*context.ParentActorID)
}

File: web/src/lib/game/context.ts
import z from 'zod'

const ContextSchema = z.object({
  source_player_ID: z.string().nullable(),

  parent_actor_ID: z.string().nullable(),
  source_actor_ID: z.string().nullable(),

  target_actor_IDs: z.array(z.string()),
  target_position_IDs: z.array(z.string()),
})

type Context = z.output<typeof ContextSchema>

function contextToString(c: Context): string {
  return `${c.parent_actor_ID}.${c.source_actor_ID}.${c.source_player_ID}.${c.target_actor_IDs.join('+')}.${c.target_position_IDs.join('+')}`
}

export { ContextSchema, contextToString }
export type { Context }
